File: flutter_plugin/ios/Classes/FlutterPlugin.swift
import Flutter
import UIKit
import AVFoundation

enum FlutterPluginChannelNames {
  static let method = "com.example.flutter_plugin/method"
  static let events = "com.example.flutter_plugin/events"
  static let messageString = "com.example.flutter_plugin/message_string"
  static let messageStandard = "com.example.flutter_plugin/message_standard"
}

enum PluginErrorCodes {
  static let ok = 0
  static let badArgs = 1
  static let notSupported = 2
  static let permissionDenied = 3
  static let permissionPermanentlyDenied = 4
  static let internalError = 5
}

final class PluginException: Error {
  init(code: Int, message: String) {
    self.code = code
    self.message = message
  }

  let code: Int
  let message: String
}

enum Args {
  static func asMap(_ any: Any?) throws -> [String: Any] {
    guard let map = any as? [String: Any] else {
      throw PluginException(code: PluginErrorCodes.badArgs, message: "Invalid arguments")
    }
    return map
  }

  static func getString(_ map: [String: Any], _ key: String) throws -> String {
    guard let value = map[key] else {
      throw PluginException(code: PluginErrorCodes.badArgs, message: "Missing \(key)")
    }
    return String(describing: value)
  }

  static func getInt(_ map: [String: Any], _ key: String) throws -> Int {
    guard let value = map[key] else {
      throw PluginException(code: PluginErrorCodes.badArgs, message: "Missing \(key)")
    }
    if let intValue = value as? Int { return intValue }
    if let number = value as? NSNumber { return number.intValue }
    if let str = value as? String, let intValue = Int(str) { return intValue }
    throw PluginException(code: PluginErrorCodes.badArgs, message: "Invalid int for \(key)")
  }

  static func getMap(_ map: [String: Any], _ key: String) throws -> [String: Any] {
    guard let value = map[key] else {
      throw PluginException(code: PluginErrorCodes.badArgs, message: "Missing \(key)")
    }
    guard let dict = value as? [String: Any] else {
      throw PluginException(code: PluginErrorCodes.badArgs, message: "Invalid map for \(key)")
    }
    return dict
  }
}

enum PluginResponse {
  static func ok(requestId: String, data: Any?) -> [String: Any?] {
    return [
      "code": PluginErrorCodes.ok,
      "message": "ok",
      "data": data,
      "requestId": requestId,
    ]
  }

  static func error(requestId: String, code: Int, message: String, data: Any? = nil) -> [String: Any?] {
    return [
      "code": code,
      "message": message,
      "data": data,
      "requestId": requestId,
    ]
  }
}

public class SwiftFlutterPluginPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  private var ticker: DispatchSourceTimer?
  private var tickCount: Int64 = 0
  private var tickIntervalMs: Int64 = 0

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = SwiftFlutterPluginPlugin()

    let method = FlutterMethodChannel(name: FlutterPluginChannelNames.method, binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: method)

    let events = FlutterEventChannel(name: FlutterPluginChannelNames.events, binaryMessenger: registrar.messenger())
    events.setStreamHandler(instance)

    let messageString = FlutterBasicMessageChannel(
      name: FlutterPluginChannelNames.messageString,
      binaryMessenger: registrar.messenger(),
      codec: FlutterStringCodec.sharedInstance()
    )
    messageString.setMessageHandler { message, reply in
      instance.handleStringMessage(message: message as? String, reply: reply)
    }

    let messageStandard = FlutterBasicMessageChannel(
      name: FlutterPluginChannelNames.messageStandard,
      binaryMessenger: registrar.messenger(),
      codec: FlutterStandardMessageCodec.sharedInstance()
    )
    messageStandard.setMessageHandler { message, reply in
      instance.handleStandardMessage(message: message, reply: reply)
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let requestId: String
    do {
      let args = try Args.asMap(call.arguments)
      requestId = (args["requestId"] as? String) ?? ""

      switch call.method {
      case "getPlatformVersion":
        result(PluginResponse.ok(requestId: requestId, data: "iOS \(UIDevice.current.systemVersion)"))
      case "getDeviceInfo":
        result(PluginResponse.ok(requestId: requestId, data: getDeviceInfo()))
      case "requestCameraPermission":
        requestCameraPermission(requestId: requestId, result: result)
      case "startTicking":
        let payload = try Args.getMap(args, "payload")
        let intervalMs = Int64(try Args.getInt(payload, "intervalMs"))
        startTicking(requestId: requestId, intervalMs: intervalMs, result: result)
      case "stopTicking":
        stopTicking()
        result(PluginResponse.ok(requestId: requestId, data: nil))
      default:
        result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.notSupported, message: "Not supported: \(call.method)"))
      }
    } catch let e as PluginException {
      let req = (try? Args.asMap(call.arguments))?["requestId"] as? String ?? ""
      result(PluginResponse.error(requestId: req, code: e.code, message: e.message))
    } catch {
      let req = (try? Args.asMap(call.arguments))?["requestId"] as? String ?? ""
      result(PluginResponse.error(requestId: req, code: PluginErrorCodes.internalError, message: "\(error)"))
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    stopTicking()
    return nil
  }

  private func getDeviceInfo() -> [String: Any?] {
    let info = Bundle.main.infoDictionary
    let versionName = info?["CFBundleShortVersionString"] as? String
    let versionCode = info?["CFBundleVersion"] as? String
    return [
      "platform": "ios",
      "model": UIDevice.current.model,
      "name": UIDevice.current.name,
      "systemVersion": UIDevice.current.systemVersion,
      "appVersionName": versionName,
      "appVersionCode": versionCode,
    ]
  }

  private func requestCameraPermission(requestId: String, result: @escaping FlutterResult) {
    let current = AVCaptureDevice.authorizationStatus(for: .video)
    switch current {
    case .authorized:
      result(PluginResponse.ok(requestId: requestId, data: ["status": "granted"]))
    case .restricted:
      result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.permissionDenied, message: "permission restricted", data: ["status": "restricted"]))
    case .denied:
      result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.permissionPermanentlyDenied, message: "permission denied", data: ["status": "permanentlyDenied"]))
    case .notDetermined:
      AVCaptureDevice.requestAccess(for: .video) { granted in
        DispatchQueue.main.async {
          let status = granted ? "granted" : "denied"
          let code = granted ? PluginErrorCodes.ok : PluginErrorCodes.permissionDenied
          let message = granted ? "ok" : "permission denied"
          let response: [String: Any?]
          if code == PluginErrorCodes.ok {
            response = PluginResponse.ok(requestId: requestId, data: ["status": status])
          } else {
            response = PluginResponse.error(requestId: requestId, code: code, message: message, data: ["status": status])
          }
          result(response)

          self.eventSink?([
            "eventName": "permission",
            "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
            "payload": [
              "permission": "camera",
              "status": status,
            ],
          ])
        }
      }
    @unknown default:
      result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.internalError, message: "unknown authorization status"))
    }
  }

  private func startTicking(requestId: String, intervalMs: Int64, result: @escaping FlutterResult) {
    guard intervalMs > 0 else {
      result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.badArgs, message: "intervalMs must be > 0"))
      return
    }
    guard eventSink != nil else {
      result(PluginResponse.error(requestId: requestId, code: PluginErrorCodes.notSupported, message: "EventChannel not listened"))
      return
    }

    stopTicking()

    tickIntervalMs = intervalMs
    tickCount = 0

    let timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
    timer.schedule(deadline: .now(), repeating: .milliseconds(Int(intervalMs)))
    timer.setEventHandler { [weak self] in
      guard let self else { return }
      self.tickCount += 1
      self.eventSink?([
        "eventName": "tick",
        "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
        "payload": [
          "count": self.tickCount,
          "intervalMs": self.tickIntervalMs,
        ],
      ])
    }
    ticker = timer
    timer.resume()
    result(PluginResponse.ok(requestId: requestId, data: nil))
  }

  private func stopTicking() {
    if let timer = ticker {
      timer.cancel()
    }
    ticker = nil
  }

  private func handleStringMessage(message: String?, reply: FlutterReply) {
    guard let text = message else {
      reply(encodeJsonString(PluginResponse.error(requestId: "", code: PluginErrorCodes.badArgs, message: "empty message")))
      return
    }

    guard let data = text.data(using: .utf8) else {
      reply(encodeJsonString(PluginResponse.error(requestId: "", code: PluginErrorCodes.badArgs, message: "invalid string")))
      return
    }

    let dictAny = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    let requestId = dictAny?["requestId"] as? String ?? ""
    let payload = dictAny?["payload"] as? [String: Any]
    let input = payload?["text"] as? String ?? ""
    let output = "iOS echo: \(input)"
    reply(encodeJsonString(PluginResponse.ok(requestId: requestId, data: output)))
  }

  private func handleStandardMessage(message: Any?, reply: FlutterReply) {
    guard let req = message as? [String: Any] else {
      reply(PluginResponse.error(requestId: "", code: PluginErrorCodes.badArgs, message: "invalid message"))
      return
    }

    let requestId = req["requestId"] as? String ?? ""
    let payload = req["payload"] as? [String: Any] ?? [:]
    let bytes = FlutterStandardTypedData(bytes: Data([1, 2, 3, 4]))
    let data: [String: Any] = [
      "platform": "ios",
      "receivedPayload": payload,
      "bytes": bytes,
      "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
    ]
    reply(PluginResponse.ok(requestId: requestId, data: data))
  }

  private func encodeJsonString(_ any: Any) -> String {
    let obj: Any
    if let dict = any as? [String: Any?] {
      obj = dict.reduce(into: [String: Any]()) { acc, item in
        acc[item.key] = item.value ?? NSNull()
      }
    } else {
      obj = any
    }

    guard let data = try? JSONSerialization.data(withJSONObject: obj),
          let str = String(data: data, encoding: .utf8) else {
      return ""
    }
    return str
  }
}

File: flutter_plugin/android/src/main/kotlin/com/example/flutter_plugin/FlutterPlugin.kt
package com.example.flutter_plugin

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.common.StringCodec
import org.json.JSONObject

open class FlutterPluginImpl :
    FlutterPlugin,
    ActivityAware,
    MethodCallHandler,
    EventChannel.StreamHandler,
    PluginRegistry.RequestPermissionsResultListener {
    private lateinit var applicationContext: Context
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var messageStringChannel: BasicMessageChannel<String>
    private lateinit var messageStandardChannel: BasicMessageChannel<Any?>

    private var eventSink: EventChannel.EventSink? = null
    private val ticker = Ticker { tickCount, timestampMs ->
        val sink = eventSink ?: return@Ticker
        val event = mapOf(
            "eventName" to "tick",
            "timestamp" to timestampMs,
            "payload" to mapOf(
                "count" to tickCount
            )
        )
        sink.success(event)
    }

    private var pendingPermissionResult: Result? = null
    private var pendingPermissionRequestId: String? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, FlutterPluginChannelNames.METHOD)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, FlutterPluginChannelNames.EVENTS)
        eventChannel.setStreamHandler(this)

        messageStringChannel = BasicMessageChannel(binding.binaryMessenger, FlutterPluginChannelNames.MESSAGE_STRING, StringCodec.INSTANCE)
        messageStringChannel.setMessageHandler { message, reply ->
            reply.reply(handleStringMessage(message))
        }

        messageStandardChannel = BasicMessageChannel(binding.binaryMessenger, FlutterPluginChannelNames.MESSAGE_STANDARD, StandardMessageCodec.INSTANCE)
        messageStandardChannel.setMessageHandler { message, reply ->
            reply.reply(handleStandardMessage(message))
        }
    }

    override fun onMethodCall(call: io.flutter.plugin.common.MethodCall, result: Result) {
        val args = call.arguments
        val requestMap = args as? Map<*, *>
        val requestId = requestMap?.get("requestId")?.toString() ?: ""

        try {
            when (call.method) {
                "getPlatformVersion" -> {
                    result.success(PluginResponse.ok(requestId, "Android ${Build.VERSION.RELEASE}"))
                }
                "getDeviceInfo" -> {
                    result.success(PluginResponse.ok(requestId, getDeviceInfo()))
                }
                "requestCameraPermission" -> {
                    handleCameraPermission(requestMap, result)
                }
                "startTicking" -> {
                    handleStartTicking(requestMap, requestId, result)
                }
                "stopTicking" -> {
                    ticker.stop()
                    result.success(PluginResponse.ok(requestId, null))
                }
                else -> {
                    result.success(PluginResponse.error(requestId, PluginErrorCodes.NOT_SUPPORTED, "Not supported: ${call.method}"))
                }
            }
        } catch (e: PluginException) {
            result.success(PluginResponse.error(requestId, e.code, e.message))
        } catch (e: Throwable) {
            result.success(PluginResponse.error(requestId, PluginErrorCodes.INTERNAL_ERROR, e.message ?: "internal_error"))
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        messageStringChannel.setMessageHandler(null)
        messageStandardChannel.setMessageHandler(null)
        ticker.stop()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeRequestPermissionsResultListener(this)
        activityBinding = null
        activity = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        ticker.stop()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode != REQUEST_CODE_CAMERA) return false
        val result = pendingPermissionResult ?: return true
        val requestId = pendingPermissionRequestId ?: ""

        pendingPermissionResult = null
        pendingPermissionRequestId = null

        val permission = permissions.firstOrNull().orEmpty()
        val granted = grantResults.firstOrNull() == PackageManager.PERMISSION_GRANTED

        val currentActivity = activity
        val status = when {
            granted -> "granted"
            currentActivity == null -> "denied"
            !currentActivity.shouldShowRequestPermissionRationale(permission) -> "permanentlyDenied"
            else -> "denied"
        }

        val responseCode = when (status) {
            "granted" -> PluginErrorCodes.OK
            "permanentlyDenied" -> PluginErrorCodes.PERMISSION_PERMANENTLY_DENIED
            else -> PluginErrorCodes.PERMISSION_DENIED
        }

        val responseMessage = when (status) {
            "granted" -> "ok"
            "permanentlyDenied" -> "permission permanently denied"
            else -> "permission denied"
        }

        val data = mapOf("status" to status)
        val response = if (responseCode == PluginErrorCodes.OK) {
            PluginResponse.ok(requestId, data)
        } else {
            PluginResponse.error(requestId, responseCode, responseMessage).toMutableMap().also {
                it["data"] = data
            }
        }

        result.success(response)

        eventSink?.success(
            mapOf(
                "eventName" to "permission",
                "timestamp" to System.currentTimeMillis(),
                "payload" to mapOf(
                    "permission" to permission,
                    "status" to status
                )
            )
        )

        return true
    }

    private fun getDeviceInfo(): Map<String, Any?> {
        val pm = applicationContext.packageManager
        val pkg = applicationContext.packageName
        val pkgInfo = try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                pm.getPackageInfo(pkg, PackageManager.PackageInfoFlags.of(0))
            } else {
                @Suppress("DEPRECATION")
                pm.getPackageInfo(pkg, 0)
            }
        } catch (_: Throwable) {
            null
        }
        val versionName = pkgInfo?.versionName
        val versionCode = pkgInfo?.let {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) it.longVersionCode else @Suppress("DEPRECATION") it.versionCode.toLong()
        }

        return mapOf(
            "platform" to "android",
            "model" to Build.MODEL,
            "manufacturer" to Build.MANUFACTURER,
            "systemVersion" to Build.VERSION.RELEASE,
            "sdkInt" to Build.VERSION.SDK_INT,
            "appVersionName" to versionName,
            "appVersionCode" to versionCode
        )
    }

    private fun handleCameraPermission(requestMap: Map<*, *>?, result: Result) {
        if (requestMap == null) throw PluginException(PluginErrorCodes.BAD_ARGS, "Invalid arguments")
        val requestId = Args.getString(requestMap, "requestId")
        val currentActivity = activity ?: run {
            result.success(PluginResponse.error(requestId, PluginErrorCodes.NOT_SUPPORTED, "No Activity attached"))
            return
        }

        if (pendingPermissionResult != null) {
            result.success(PluginResponse.error(requestId, PluginErrorCodes.BAD_ARGS, "Permission request in progress"))
            return
        }

        val granted = currentActivity.checkSelfPermission(Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED
        if (granted) {
            result.success(PluginResponse.ok(requestId, mapOf("status" to "granted")))
            return
        }

        pendingPermissionResult = result
        pendingPermissionRequestId = requestId
        currentActivity.requestPermissions(arrayOf(Manifest.permission.CAMERA), REQUEST_CODE_CAMERA)
    }

    private fun handleStartTicking(
        requestMap: Map<*, *>?,
        requestId: String,
        result: Result
    ) {
        if (requestMap == null) throw PluginException(PluginErrorCodes.BAD_ARGS, "Invalid arguments")
        if (eventSink == null) {
            result.success(PluginResponse.error(requestId, PluginErrorCodes.NOT_SUPPORTED, "EventChannel not listened"))
            return
        }
        val payload = Args.getMap(requestMap, "payload")
        val intervalMs = Args.getInt(payload, "intervalMs").toLong()
        if (intervalMs <= 0) {
            result.success(PluginResponse.error(requestId, PluginErrorCodes.BAD_ARGS, "intervalMs must be > 0"))
            return
        }
        ticker.start(intervalMs)
        result.success(PluginResponse.ok(requestId, null))
    }

    private fun handleStringMessage(message: String?): String {
        val json = try {
            JSONObject(message ?: "")
        } catch (_: Throwable) {
            return JSONObject()
                .put("code", PluginErrorCodes.BAD_ARGS)
                .put("message", "Invalid JSON")
                .put("data", JSONObject.NULL)
                .put("requestId", "")
                .toString()
        }

        val requestId = json.optString("requestId", "")
        val payload = json.optJSONObject("payload")
        val text = payload?.optString("text", "") ?: ""
        val echoed = "Android echo: $text"
        return JSONObject(PluginResponse.ok(requestId, echoed)).toString()
    }

    @Suppress("UNCHECKED_CAST")
    private fun handleStandardMessage(message: Any?): Any? {
        val requestMap = message as? Map<*, *> ?: return PluginResponse.error("", PluginErrorCodes.BAD_ARGS, "Invalid message")
        val requestId = requestMap["requestId"]?.toString() ?: ""
        return try {
            val payload = Args.getMap(requestMap, "payload")
            val data = mapOf(
                "platform" to "android",
                "receivedPayload" to payload,
                "bytes" to byteArrayOf(1, 2, 3, 4),
                "timestamp" to System.currentTimeMillis()
            )
            PluginResponse.ok(requestId, data)
        } catch (e: PluginException) {
            PluginResponse.error(requestId, e.code, e.message)
        } catch (e: Throwable) {
            PluginResponse.error(requestId, PluginErrorCodes.INTERNAL_ERROR, e.message ?: "internal_error")
        }
    }

    private companion object {
        const val REQUEST_CODE_CAMERA = 23101
    }
}
